src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Constants.cs

```diff
@@ -28,5 +30,39 @@ internal static class Constants
         /// hashing ValueTypeDefaultComparerFrozenDictionary/Set.
         /// </remarks>
         public const int MaxItemsInSmallValueTypeFrozenCollection = 10;
+
+        /// <summary>
+        /// Whether the <typeparamref name="T"/> is known to implement <see cref="IComparable{T}"/> safely and efficiently,
+        /// such that its comparison operations should be used in searching for types in small collections.
+        /// </summary>
+        /// <remarks>
+        /// This does not automatically return true for any type that implements <see cref="IComparable{T}"/>.
+        /// Doing so leads to problems for container types (e.g. ValueTuple{T1, T2}) where the
+        /// container implements <see cref="IComparable{T}"/> to delegate to its contained items' implementation
+        /// but those then don't provide such support.
+        /// </remarks>
+        public static bool IsKnownComparable<T>() =>
+            // This list covers all of the IComparable<T> value types in Corelib that aren't containers (like ValueTuple).
+            typeof(T).IsPrimitive ||
```

Comment:
How can we make sure that this list stays up to date as new "simple value types" are being added to System.Private.CoreLib? Could this be extracted to a shared internal method (and possibly used elsewhere as well?)

Reply:
I'm not sure it really matters.  Most of these are rarely if ever used as keys in a dictionary, and even if they are, the gains that come from using the implementation based on this is only marginal in the common case.  I think it's fine to leave it as-is, and then update it adhoc should we find a need to include something else.

Reply:
Should we incorporate checks for `Nullable<T>` instances of these types?

Reply:
`Nullable<T>` isn't `IComparable<Nullable<T>>`.  If we wanted to special-case `Nullable<T>`, I think we'd want to do so with a dedicated implementation.